lbry/lbry/extras/daemon/Components.py

```diff
@@ -136,86 +146,60 @@ def component(self):
                 log.warning("s3 appears to have corrupted header")
                 return
             final_size_after_download = response.content_length + local_header_size
-            write_mode = "wb"
             if local_header_size > 0:
                 log.info("Resuming download of %i bytes from s3", response.content_length)
-                write_mode = "a+b"
-            with open(self.headers_file, write_mode) as fd:
-                while True:
-                    chunk = await response.content.read(512)
-                    if not chunk:
-                        break
-                    fd.write(chunk)
-                    self._headers_progress_percent = math.ceil(
-                        float(fd.tell()) / float(final_size_after_download) * 100
-                    )
-            log.info("fetched headers from s3, now verifying integrity after download.")
-            self._check_header_file_integrity()
-
-    def local_header_file_height(self):
-        return max((self.local_header_file_size() / HEADER_SIZE) - 1, 0)
+            buffer, header_size = b'', self.headers.header_size
+            async for chunk in response.content.iter_any():
+                chunk = buffer + chunk
+                remaining = len(chunk) % header_size
+                chunk, buffer = chunk[:-remaining], bytes(chunk[-remaining:])
+                if not chunk:
+                    continue
+                if not await self.headers.connect(len(self.headers), chunk):
+                    log.warning("Error connecting downloaded headers from at %s.", self.headers.height)
+                    return
+                self._headers_progress_percent = math.ceil(
+                    float(self.headers.bytes_size) / float(final_size_after_download) * 100
+                )
 
     def local_header_file_size(self):
         if os.path.isfile(self.headers_file):
             return os.stat(self.headers_file).st_size
         return 0
 
-    async def get_remote_height(self):
-        ledger = SimpleNamespace()
-        ledger.config = {
-            'default_servers': self.conf.lbryum_servers,
-            'data_path': self.conf.wallet_dir
-        }
-        net = Network(ledger)
-        first_connection = net.on_connected.first
-        asyncio.ensure_future(net.start())  # TODO: SKETCHY! it might be trapping a CancelledError and not raising it
-        await first_connection
-        remote_height = await net.get_server_height()
-        await net.stop()
-        return remote_height
+    async def get_download_height(self):
+        async with utils.aiohttp_request('HEAD', HEADERS_URL) as response:
+            if response.status != 200:
+                log.warning("Header download error: %s", response.status)
+                return 0
+            return response.content_length // HEADER_SIZE
```

Comment:
Perhaps add the `try`/`except` inside of this function instead of in `should_download_headers_from_s3` at line 185, that would make the logic in `should_download_headers_from_s3` a bit simpler.